esrally/track/params.py

```diff
@@ -953,39 +953,54 @@ def create_readers(
-    for group, corpus in enumerate(corpora[(start_client_index + mod) % len(corpora)] for mod in range(len(corpora))):
-        for entry, docs in enumerate(corpus.documents):
+    reordered_corpora = corpora[start_client_index:] + corpora[:start_client_index]
+    for corpus in reordered_corpora:
+        readers.append(collections.deque())
+        for docs in corpus.documents:
             offset, num_docs, num_lines = bounds(
                 docs.number_of_documents, start_client_index, end_client_index, num_clients, docs.includes_action_and_meta_data
             )
-            if num_docs > 0:
-                target = f"{docs.target_index}/{docs.target_type}" if docs.target_index else "/"
-                if docs.target_data_stream:
-                    target = docs.target_data_stream
-                logger.debug(
-                    "Task-relative clients at index [%d-%d] will bulk index [%d] docs starting from line offset [%d] for [%s] "
-                    "from corpus [%s].",
-                    start_client_index,
-                    end_client_index,
-                    num_docs,
-                    offset,
-                    target,
-                    corpus.name,
-                )
-                readers[len(corpora) * entry + group] = create_reader(
-                    docs, offset, num_lines, num_docs, batch_size, bulk_size, id_conflicts, conflict_probability, on_conflict, recency
-                )
-            else:
+            if num_docs == 0:
                 logger.debug(
                     "Task-relative clients at index [%d-%d] skip [%s] (no documents to read).",
                     start_client_index,
                     end_client_index,
                     corpus.name,
                 )
-    return readers
+                continue
+
+            target = f"{docs.target_index}/{docs.target_type}" if docs.target_index else "/"
+            if docs.target_data_stream:
+                target = docs.target_data_stream
+            logger.debug(
+                "Task-relative clients at index [%d-%d] will bulk index [%d] docs starting from line offset [%d] for [%s] "
+                "from corpus [%s].",
+                start_client_index,
+                end_client_index,
+                num_docs,
+                offset,
+                target,
+                corpus.name,
+            )
+            reader = create_reader(
+                docs, offset, num_lines, num_docs, batch_size, bulk_size, id_conflicts, conflict_probability, on_conflict, recency
+            )
+            readers[-1].append(reader)
+            total_readers += 1
+
+    # Instead of reading all files from the first corpus, and then all files from the
+    # second corpus, etc. we want to read the first file of all corpora, then the second
+    # file, etc.
+    staggered_readers = []
+    while total_readers > 0:
+        for reader_queue in readers:
+            if reader_queue:
```

Comment:
I *think* this is only needed because we eagerly add the dequeue to `readers`? If we'd only add it when there are actually documents that need to be read, we could get rid of this if.

Reply:
No, actually this is needed because all queues don't necessarily have the same length. Say I have the following readers list of deques: [[doc1], [doc1, doc2]] and total_reads = 3. After one iteration I will have the following readers: [[], [doc2]] and total_reads = 1. I need to skip the empty deque.

I tried to explain this in [`8f05ed2` (#1659)](https://github.com/elastic/rally/pull/1659/commits/8f05ed285622ebc96940753e07ec418890ca16b0).